Implement stake-based quorum balancer

AndrejMitrovic opened the pull request:
This is the second part of #240, and it implements #963.

There may be further plans to refine the algorithm:

- Take the preimages into account. This should be fairly simple, we will just add it as a second parameter to `getGenerator()`. However there are no mechanisms in place right now to retrieve preimages of all validators at a specific height.
- Take the history of the validator's performance into account. This could be a form of "score", a combination of the node's stake + historical voting data in the blockchain.

Fixes #963

AndrejMitrovic commented:
@TrustHenry help me I have a Windows linker error (your favorite)!

```
agora-unittests.obj : error LNK2019: unresolved external symbol "public: static class std::shared_ptr<class stellar::QuorumIntersectionChecker *> __cdecl stellar::QuorumIntersectionChecker::create(struct std::unordered_map<struct stellar::PublicKey,class std::shared_ptr<struct stellar::SCPQuorumSet>,struct std::hash<struct stellar::PublicKey>,struct std::equal_to<struct stellar::PublicKey>,class std::allocator<struct std::pair<struct stellar::PublicKey const ,class std::shared_ptr<struct stellar::SCPQuorumSet> > > > const &)" (?create@QuorumIntersectionChecker@stellar@@SA?AV?$shared_ptr@PEAVQuorumIntersectionChecker@stellar@@@std@@AEBU?$unordered_map@UPublicKey@stellar@@V?$shared_ptr@USCPQuorumSet@stellar@@@std@@U?$hash@UPublicKey@stellar@@@4@U?$equal_to@UPublicKey@stellar@@@4@V?$allocator@U?$pair@$$CBUPublicKey@stellar@@V?$shared_ptr@USCPQuorumSet@stellar@@@std@@@std@@@4@@4@@Z) referenced in function _D5agora9consensus6Quorum22verifyQuorumsIntersectFHSQBy6common6crypto3Key9PublicKeySQDeQBg5Types12QuorumConfigZv
```

TrustHenry commented:
> um.. Windows linker error

pragma(mangle, `?create@ ....

Wouldn't this be a temporary solution?

AndrejMitrovic commented:
> Could not deserialize value

Ok it seems that the ballot timer erroring out ("moved to an invalid state") is some kind of memory corruption bug. I'm looking into it.

AndrejMitrovic commented:
I found the root cause of the memory corruption bug (and it would happen in master too), this is the reason why we had "moved to a bad state" errors frequently.

I added a fixup commit, but it's not very efficient - I just wanted to unblock my work. I could split it into another PR too.

AndrejMitrovic commented:
I will split off some commits from this PR.

AndrejMitrovic committed:
Implement stake-based quorum balancer

Nodes which have a higher stake compared
to the rest of the network will have a
higher chance of being included in other
node's quorum sets.

There are two constants which alter the
generator's behavior:

- MAX_NODES_IN_QUORUM
  Maximum number of nodes to include in a quorum.

- THRESHOLD
  Threshold to use for the quorum configuration (e.g. 0.80 = 80%).

The algorithm uses a random number generator to shuffle
the quorum node selection. The public key of a node is used
to initialize the RNG. In a later edition we will also
combine it with the concatenation of all preimages to
make the shuffling less predictable.

Fixes: #963

AndrejMitrovic committed:
Move verifyQuorumSanity to end and version it

It should only be available in unitests.

AndrejMitrovic committed:
Add ability to check if the Validator is enrolled

This is temporary support code until #798 is resolved.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+        quorum.nodes ~= stakes[idx].key;
+        added[idx] = true;  // mark used
+
+        if (quorum.nodes.length >= stakes.length + 1)  // our stake not included
```
I think this condition could be moved inside the `while` condition.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
     return quorum;
 }
 
+version (unittest)
```
This should be moved somewhere else so it doesn't interfere with ddoc comments.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+    return WK.Keys.byRange.take(count).map!(kp => kp.address).array;
+}
+
+/// 1 node
```
Only one of these unittests should be ddoc-documented, the rest doesn't have to be.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+        [111, 165, 189, 80, 37, 5, 16, 194, 39, 214, 156, 169, 235, 221, 21, 126];
+
+    ubyte[ulong.sizeof] short_hash;
+    crypto_shorthash(short_hash.ptr, hash[].ptr, 64, PublicKey.ptr);
```
```suggestion
    crypto_shorthash(short_hash.ptr, hash[].ptr, hash[].length, PublicKey.ptr);
```

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+    import std.bitmanip;
+
+    // generated once with 'crypto_shorthash_keygen'
+    static immutable ubyte[crypto_shorthash_KEYBYTES] PublicKey =
```
`PublicKey` is a bad name for this. We should just name it `iv` as a kind of initialization vector.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+    initializer for the RNG engine.
+
+    Using the Mersene Twister 19937 64-bit random number generator.
+    The public key is reduced to a short hash of 8 bytes,
```
```suggestion
    The public key is reduced to a short hash of 8 bytes
```

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+private auto getGenerator (PublicKey key)
+{
+    Mt19937_64 gen;
+    auto hash = hashFull(key[]);  // key is only 32 bytes, hash is 64
```
So why not just change `toShortHash` to take a 32 byte value?

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+        filter = the node's own key should be filtered here
         utxo_keys = the list of enrollments' UTXO keys
-        finder = UTXO finder delegate
+        finder = utxo to stake & public key lookup delegate
```
```suggestion
        finder = lookup delegate to find the public key & stake
```

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
 
-private Set!PublicKey getPublicKeys (in Hash[] utxo_keys, in UTXOFinder finder)
+private NodeStake[] buildStakesDescending (const ref PublicKey filter,
+    in Hash[] utxo_keys, UTXOFinder finder)
```
```suggestion
    in Hash[] utxo_keys, UTXOFinder finder) nothrow
```

Maybe?

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
 
     Returns:
-        a tuple of (Enrollment[], UTXOFinder)
+        the map of node => quorum set inclusion counts
```
This documentation is wrong.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+*******************************************************************************/
+
+version (unittest)
+private size_t[] countInclusions (QuorumConfig[PublicKey] quorums,
```
```suggestion
private size_t[] countNodeInclusions (QuorumConfig[PublicKey] quorums,
```

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,181 +13,476 @@
+    another node's quorum set.
+
+    Returns:
+        the map of node => quorum set inclusion counts
```
It's not a map anymore, but an array. Document why.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -112,20 +112,34 @@ extern(D):
         Set or update the quorum configuration
 
+        The node additionally takes the list of all other validator's quorum
+        configurations and saves them for lookup in getQSet().
```
```suggestion
        configurations and saves them for later lookup by getQSet().
```

AndrejMitrovic commented:
The commit message should talk a bit more how the algorithm works.

AndrejMitrovic commented:
Some design comments:

- I think we can further iterate and alter the design of the algorithm as we move along and learn new things. It doesn't have to be perfect from the start.
- I thought about the network communication overhead costs. But I came to the conclusion that the quorum design alone won't have that much of an impact on latency. For example if node X is included in 10 other node's quorums, it doesn't necessarily mean those nodes must directly communicate with node X. They are only interested in there being a signed message from X, and that message may be propagated by other nodes. It should the the role of the network manager to find the most efficient way to get a message to a particular node.
- I also thought about whether "past performance" of a Validator should be taken into account in quorum balancing in addition to the node's stake. If for example there is a validator with a lot of stake but which fails / refuses to sign many blocks then the quorum balancer should punish that node by lowering its score. But thinking about this again, this becomes a little bit hard to do because we don't really have "identity", it is tied to the public key. A node operator can easily generate a new key, use a different UTXO, etc..
- I am not sure about the edge-cases yet. For example, probability could make it possible a validator is not included in any quorum (if its stake is very tiny compared to the rest of the network). I should try to write a test-case for this.

Anyway my theory is that we will refine this many more times before Coinnet.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -522,6 +522,26 @@ public class EnrollmentManager
     }
 
+    /// Returns: true if this validator is currently enrolled
+    public bool isEnrolled (UTXOFinder finder) nothrow @safe
```
```suggestion
    public bool isEnrolled (scope UTXOFinder finder) nothrow @safe
```

Geod24 commented on source/agora/consensus/Quorum.d:
```diff
@@ -13,180 +13,478 @@
+unittest
+{
+    auto keys = getKeys(128);
+    auto amounts = Amount(400000000000uL).repeat(128).array;
+    amounts[0] = amounts[$ - 1] = Amount(400000000000uL * 126);
```
Please use `_` for big numbers

Geod24 reviewed (approved):
LGTM overall. Some nits but nothing blocking. As you said, we'll iterate over this quite a few times.
